src/alpakatest/plugin-Test1/alpaka/alpakaAlgo1.cc

```diff
@@ -106,29 +316,67 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE {
                                                                  alpaka::mem::view::getPtrNative(d_c_buf),
                                                                  NUM_VALUES));
 
-#ifdef TODO
-    vectorAdd<<<blocksPerGrid, threadsPerBlock, 0, stream>>>(d_a.get(), d_b.get(), d_c.get(), NUM_VALUES);
-
-    auto d_ma = cms::cuda::make_device_unique<float[]>(NUM_VALUES * NUM_VALUES, stream);
-    auto d_mb = cms::cuda::make_device_unique<float[]>(NUM_VALUES * NUM_VALUES, stream);
-    auto d_mc = cms::cuda::make_device_unique<float[]>(NUM_VALUES * NUM_VALUES, stream);
-    dim3 threadsPerBlock3{NUM_VALUES, NUM_VALUES};
-    dim3 blocksPerGrid3{1, 1};
-    if (NUM_VALUES * NUM_VALUES > 32) {
-      threadsPerBlock3.x = 32;
-      threadsPerBlock3.y = 32;
-      blocksPerGrid3.x = ceil(double(NUM_VALUES) / double(threadsPerBlock3.x));
-      blocksPerGrid3.y = ceil(double(NUM_VALUES) / double(threadsPerBlock3.y));
-    }
-    vectorProd<<<blocksPerGrid3, threadsPerBlock3, 0, stream>>>(d_a.get(), d_b.get(), d_ma.get(), NUM_VALUES);
-    vectorProd<<<blocksPerGrid3, threadsPerBlock3, 0, stream>>>(d_a.get(), d_c.get(), d_mb.get(), NUM_VALUES);
-    matrixMul<<<blocksPerGrid3, threadsPerBlock3, 0, stream>>>(d_ma.get(), d_mb.get(), d_mc.get(), NUM_VALUES);
+    // Prepare 2D workDiv
+    Vec2 elementsPerThread2(1u, 1u);
+    const unsigned int threadsPerBlockSide = (NUM_VALUES < 32 ? NUM_VALUES : 32u);
+    Vec2 threadsPerBlock2(threadsPerBlockSide, threadsPerBlockSide);
+#if defined ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLED || ALPAKA_ACC_CPU_B_TBB_T_SEQ_ENABLED || \
+  ALPAKA_ACC_CPU_B_OMP2_T_SEQ_ENABLED || ALPAKA_ACC_CPU_BT_OMP4_ENABLED
+    // on the GPU, run with 32 threads in parallel per block, each looking at a single element
+    // on the CPU, run serially with a single thread per block, over 32 elements
+    std::swap(threadsPerBlock2, elementsPerThread2);
+#endif
+    const unsigned int blocksPerGridSide = (NUM_VALUES <= 32 ? 1 : std::ceil(NUM_VALUES / 32.));
+    const Vec2 blocksPerGrid2(blocksPerGridSide, blocksPerGridSide);
+    const WorkDiv2 workDiv2(blocksPerGrid2, threadsPerBlock2, elementsPerThread2);
 
-    matrixMulVector<<<blocksPerGrid, threadsPerBlock, 0, stream>>>(d_mc.get(), d_b.get(), d_c.get(), NUM_VALUES);
 
-    return d_a;
-#endif
+    // Device data
+    const Vec2 sizeSquare(NUM_VALUES, NUM_VALUES);
+    auto d_ma_buf = alpaka::mem::buf::alloc<float, Idx>(device, sizeSquare);
+    auto d_mb_buf = alpaka::mem::buf::alloc<float, Idx>(device, sizeSquare);
+    auto d_mc_buf = alpaka::mem::buf::alloc<float, Idx>(device, sizeSquare);
 
+
+    // VECTOR MULTIPLICATION
+    alpaka::wait::wait(queue);
```

Comment:
Same here, wanted to reproduce the implicit synchronization from cudatest in https://github.com/ghugo83/pixeltrack-standalone/blob/master/src/cudatest/plugin-Test1/gpuAlgo1.cu#L89 
Will remove

Reply:
> Same here, wanted to reproduce the implicit synchronization in cudatest in https://github.com/ghugo83/pixeltrack-standalone/blob/master/src/cudatest/plugin-Test1/gpuAlgo1.cu#L89

I'm a bit confused because in CUDA the kernel calls are asynchronous. In `cudatest` case the synchronization points are somewhat implicit (use of same CUDA stream, use of `cudaStreamWaitEvent()` if reading data produced in a different stream, use of ExternalWork).